CSC8503CoreClasses/PhysicsSystem.cpp

```diff
@@ -220,51 +221,138 @@ In tutorial 5, we start determining the correct response to a collision,
+float PhysicsSystem::ImpulseResolveCollision(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p) const {
 	PhysicsObject* physA = a.GetPhysicsObject();
 	PhysicsObject* physB = b.GetPhysicsObject();
 
-	Transform& transformA = a.GetTransform();
-	Transform& transformB = b.GetTransform();
-
 	float totalMass = physA->GetInverseMass() + physB->GetInverseMass();
 
 	// both objects are static
 	if (totalMass == 0)
-		return;
-
-	transformA.SetPosition(transformA.GetPosition() - (p.normal * p.penetration * (physA->GetInverseMass() / totalMass)));
-	transformB.SetPosition(transformB.GetPosition() + (p.normal * p.penetration * (physB->GetInverseMass() / totalMass)));
+		return 0;
 
+	SeperateObjects(a, b, p, totalMass);
+	
+	// local collision points
 	Vector3 relativeA = p.localA;
 	Vector3 relativeB = p.localB;
-
-	Vector3 angVelocityA = Vector3::Cross(physA->GetAngularVelocity(), relativeA);
-	Vector3 angVelocityB = Vector3::Cross(physB->GetAngularVelocity(), relativeB);
-
-	Vector3 fullVelocityA = physA->GetLinearVelocity() + angVelocityA;
-	Vector3 fullVelocityB = physB->GetLinearVelocity() + angVelocityB;
-
-	Vector3 contactVelocity = fullVelocityB - fullVelocityA;
+	
+	Vector3 contactVelocity = CalculateCollisionVelocity(relativeA, relativeB, physA, physB);
 
 	float impulseForce = Vector3::Dot(contactVelocity, p.normal);
-
-	// calculate inertia
-	Vector3 inertiaA = Vector3::Cross(physA->GetInertiaTensor() * Vector3::Cross(relativeA, p.normal), relativeA);
-	Vector3 inertiaB = Vector3::Cross(physB->GetInertiaTensor() * Vector3::Cross(relativeB, p.normal), relativeB);
-	float angularEffect = Vector3::Dot(inertiaA + inertiaB, p.normal);
+	float angularEffect = CalculateInertia(physA, physB,  relativeA,  relativeB,  p.normal);
 
 	float cRestitution = 0.66f; // loss of kinetic energy
 
 	float j = (-(1.0f + cRestitution) * impulseForce) / (totalMass + angularEffect);
-
-	Vector3 fullImpulse = p.normal * j;
+	Vector3 fullImpulse = p.normal * j * physA->GetElasticity();
 
 	// apply impulse in opposite directions for collision responce
 	physA->ApplyLinearImpulse(-fullImpulse);
 	physB->ApplyLinearImpulse(fullImpulse);
 	physA->ApplyAngularImpulse(Vector3::Cross(relativeA, -fullImpulse));
 	physB->ApplyAngularImpulse(Vector3::Cross(relativeB, fullImpulse));
+
+	return j;
+}
+
+void PhysicsSystem::FrictionImpulse(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p, float j) const {
+	PhysicsObject* physA = a.GetPhysicsObject();
+	PhysicsObject* physB = b.GetPhysicsObject();
+	float totalMass = physA->GetInverseMass() + physB->GetInverseMass();
+	Vector3 relativeA = p.localA;
+	Vector3 relativeB = p.localB;
+
+	Vector3 contactVelocity;
+
+	contactVelocity = (GetIsCapsule(a) || GetIsCapsule(b) ? contactVelocity = physB->GetLinearVelocity() - physA->GetLinearVelocity() : contactVelocity = CalculateCollisionVelocity(relativeA, relativeB, physA, physB));
```

Comment:
This took a little bit for me to parse: I think you don't need the two 'contactVelocity = ' re-statements in a ternary. So it should read:
contactVelocity = (GetIsCapsule(a) || GetIsCapsule(b)) ? physB->GetLinearVelocity() - physA->GetLinearVelocity : 
                              CalculateCollisionVelocity(relativeA, relativeB, physA, physB);

Which may well still be a little confusing! If-else might be clearer here, but that's up to you.